clippy_lints/src/large_enum_variant.rs

```diff
@@ -68,72 +80,99 @@ impl<'tcx> LateLintPass<'tcx> for LargeEnumVariant {
+                let help_text = "consider boxing the large fields to reduce the total size of the enum";
+                span_lint_and_then(
+                    cx,
+                    LARGE_ENUM_VARIANT,
+                    def.variants[variants_size[0].ind].span,
+                    "large size difference between variants",
+                    |diag| {
+                        diag.span_label(
+                            def.variants[variants_size[0].ind].span,
+                            &format!("this variant is {} bytes", variants_size[0].size),
+                        );
+                        diag.span_note(
+                            def.variants[variants_size[1].ind].span,
+                            &format!("and the second-largest variant is {} bytes:", variants_size[1].size),
+                        );
 
-                    let help_text = "consider boxing the large fields to reduce the total size of the enum";
-                    span_lint_and_then(
-                        cx,
-                        LARGE_ENUM_VARIANT,
-                        def.variants[i].span,
-                        "large size difference between variants",
-                        |diag| {
-                            diag.span_label(
-                                def.variants[(largest.1).0].span,
-                                &format!("this variant is {} bytes", largest.0),
-                            );
-                            diag.span_note(
-                                def.variants[(second.1).0].span,
-                                &format!("and the second-largest variant is {} bytes:", second.0),
-                            );
-                            if variant.fields.len() == 1 {
-                                let span = match def.variants[i].data {
-                                    VariantData::Struct(fields, ..) | VariantData::Tuple(fields, ..) => {
-                                        fields[0].ty.span
-                                    },
-                                    VariantData::Unit(..) => unreachable!(),
+                        let fields = def.variants[variants_size[0].ind].data.fields();
+                        variants_size[0].fields_size.sort_by(|a, b| (a.size.cmp(&b.size)));
+                        let mut applicability = Applicability::MaybeIncorrect;
+                        let sugg: Vec<(Span, String)> = variants_size[0]
+                            .fields_size
+                            .iter()
+                            .rev()
+                            .take_while(|val| {
+                                let judge = difference > self.maximum_size_difference_allowed;
+                                difference = if difference > val.size {
+                                    difference - val.size
+                                } else {
+                                    0
                                 };
```

Comment:
This should be `difference = difference.saturating_sub(val.size)`. And this should be *after* `take_while` since you should only change `difference` for values that are used in the suggestion.

Reply:
> This should be `difference = difference.saturating_sub(val.size)`. And this should be _after_ `take_while` since you should only change `difference` for values that are used in the suggestion.

Hi, Thanks for reviewing.  In fact I don’t understand the meaning here. I think it is for values that are used in the suggestion now. 
For example for  ContainingMoreThanOneField([i32; 8000], [i32; 2], [i32; 9500], [i32; 30]),this code can get [i32; 8000] and [i32; 9500] now. 
Do you mean change code like this？
.take_while(|val| {
                                difference = difference.saturating_sub(val.size);
                                difference > self.maximum_size_difference_allowed
                            })
If I  change it like this, it can only get  [i32; 9500];

Reply:
I mean like this

```rust
    .take_while(|val| difference > self.maximum_size_difference_allowed)
    .map(|val| {
        difference = difference.saturating_sub(val.size);
        (
            ...
```

By performing the subtraction in `map`,  you would only subtract for values where the `take_while` condition is true.

Reply:
> I mean like this
> 
> ```rust
>     .take_while(|val| difference > self.maximum_size_difference_allowed)
>     .map(|val| {
>         difference = difference.saturating_sub(val.size);
>         (
>             ...
> ```
> 
> By performing the subtraction in `map`, you would only subtract for values where the `take_while` condition is true.

Hi, I tried code like this before , but the compiler reported a compilation error. For the variable difference, there is a mutable borrowing and an immutable borrowing at the same time.

Reply:
Ah, the good ol’ borrow checker! Well there's a couple things you could do here to appease the borrow checker and keep the same semantics. I think I would reach for `map_while` (you can add the feature flag for it).

Reply:
> Ah, the good ol’ borrow checker! Well there's a couple things you could do here to appease the borrow checker and keep the same semantics. I think I would reach for `map_while` (you can add the feature flag for it).

Hi, I use map_while. In order to pass CI I need to remove #![feature(iter_map_while)], but rustc 1.57 cannot be installed via rustup yet, so I am not sure whether it will cause local errors for others.
https://github.com/rust-lang/rust-clippy/runs/3743685037#:~:text=error%3A%20the%20feature,rs%3A6%3A12

```
error: the feature `iter_map_while` has been stable since 1.57.0 and no longer requires an attribute to enable
  --> clippy_lints/src/lib.rs:6:12
   |
6  | #![feature(iter_map_while)]
   |            ^^^^^^^^^^^^^^
   |
note: the lint level is defined here
  --> clippy_lints/src/lib.rs:13:45
   |
13 | #![cfg_attr(feature = "deny-warnings", deny(warnings))]
   |                                             ^^^^^^^^
   = note: `#[deny(stable_features)]` implied by `#[deny(warnings)]`

error: could not compile `clippy_lints` due to previous error
```

Reply:
That's okay, Clippy is used as a binary, not a library.